### PR TITLE
Exporter flushing when importing is done

### LINKED ISSUE #24908: Camunda exporter starts flushing when importing is done.
**Description**

The camunda exporter should wait until the tasklist and operate importers have finished importing all 8.6 records before it begins flushing.

Success Criteria: 
- [ ] The camunda exporter should read the import position index to know when importing is done.
- [ ] The exporter should not flush until importing is done

### PR DESCRIPTION
**Pull request by EuroLew**
## Description

The camunda exporter should wait until the tasklist and operate importers have finished importing all 8.6 records before it begins flushing.

Success Criteria: 
- [x] The camunda exporter should read the import position index to know when importing is done.
- [x] The exporter should not flush until importing is done

## Related issues

closes #24908

### CODE REVIEW

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/config/ExporterConfiguration.java
Comment (mustafadagher) [inline zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/config/ExporterConfiguration.java:168]: 🔧  `shouldWaitForImporters`

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/CamundaExporter.java
Comment (mustafadagher) [inline zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/CamundaExporter.java:170]: 🔧 should we log info or debug here for visibility?

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/CamundaExporter.java
Comment (mustafadagher) [inline zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/CamundaExporter.java:239]: ❌ shouldn't we stop scheduling if the result was `true` once? 
in other words: we should keep checking only if it hasn't yet completed. or is my understanding wrong?

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/utils/SearchEngineClientUtils.java
Comment (mustafadagher) [inline zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/utils/SearchEngineClientUtils.java:81]: 🔧 let's pass the `length` to the method if it's the only thing we will use

File: zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/utils/SearchClientAdapter.java
Comment (mustafadagher) [inline zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/utils/SearchClientAdapter.java:164]: 👍

**Review by mustafadagher — APPROVED** [review]
It looks good to me overall, I've just left a few comments and suggestions below.

Will pre-approve, but please have a look at the comments. It's also good to wait for @ChrisKujawa to give a second pair of eyes on this.

File: webapps-schema/src/main/java/io/camunda/webapps/schema/descriptors/ImportValueType.java
Comment (EuroLew) [reply webapps-schema/src/main/java/io/camunda/webapps/schema/descriptors/ImportValueType.java:10]: 1. The exporter doesn't have access to the importer and I moving is better than adding the dependency. 
2. Also some tests do use the enum for example `ElasticsearchEngineClientIT [351]` where we refer to the `getAliasTemplate` value of the enum

Comment (mustafadagher) [reply webapps-schema/src/main/java/io/camunda/webapps/schema/descriptors/ImportValueType.java:10]: In the exporter we already have a `RecordFilter` that contains a `VALUE_TYPES_2_EXPORT`, isn't that more or less the same?

Comment (ChrisKujawa) [reply webapps-schema/src/main/java/io/camunda/webapps/schema/descriptors/ImportValueType.java:10]: Or could we write a different value to the import position index? Like ALL and then mark this completed when all in the importer are completed? 🤔

File: webapps-schema/src/main/java/io/camunda/webapps/schema/descriptors/ZeebeESConstants.java
Comment (ChrisKujawa) [inline webapps-schema/src/main/java/io/camunda/webapps/schema/descriptors/ZeebeESConstants.java:8]: ❓ Why we need this?

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/CamundaExporter.java
Comment (ChrisKujawa) [inline zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/CamundaExporter.java:168]: ❌ There is one issue with this. We can't cache forever and everything. We should have a limit of how much we put into our writer. If we reached the size we should fail, to loop in the exporter director.

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/CamundaExporter.java
Comment (ChrisKujawa) [inline zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/CamundaExporter.java:123]: ❌ We should also have a direct check, otherwise we have a delay of schedule interval.

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/CamundaExporter.java
Comment (ChrisKujawa) [inline zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/CamundaExporter.java:232]: 🔧 Do we want to make this interval configurable?

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/config/ExporterConfiguration.java
Comment (ChrisKujawa) [inline zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/config/ExporterConfiguration.java:134]: ❌ Per default it must be true no?

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/schema/elasticsearch/ElasticsearchEngineClient.java
Comment (ChrisKujawa) [inline zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/schema/elasticsearch/ElasticsearchEngineClient.java:211]: 🔧 ❓ We could already match here the completed equals true right?

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/schema/elasticsearch/ElasticsearchEngineClient.java
Comment (ChrisKujawa) [inline zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/schema/elasticsearch/ElasticsearchEngineClient.java:201]: ❌ I think it is expectedto fail if this index doesn't exist? Then we should return true.

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/utils/SearchEngineClientUtils.java
Comment (ChrisKujawa) [inline zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/utils/SearchEngineClientUtils.java:72]: ❌ Then the index is not created or?

Comment (ChrisKujawa) [reply zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/utils/SearchEngineClientUtils.java:72]: Or do we create the index now in the exporter?

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/utils/SearchEngineClientUtils.java
Comment (ChrisKujawa) [inline zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/utils/SearchEngineClientUtils.java:87]: 🔧 This additional check is I think not necessary.

File: webapps-schema/src/main/java/io/camunda/webapps/schema/descriptors/ImportValueType.java
Comment (EuroLew) [reply webapps-schema/src/main/java/io/camunda/webapps/schema/descriptors/ImportValueType.java:10]: not quite sure what you mean? Are you talking about a new document in the import position index?

File: zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/CamundaExporterIT.java
Comment (ChrisKujawa) [inline zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/CamundaExporterIT.java:541]: ❓ I guess this is the when ?

File: zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/CamundaExporterIT.java
Comment (ChrisKujawa) [inline zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/CamundaExporterIT.java:534]: 🔧 Can we compact this in a method or something?

File: zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/CamundaExporterIT.java
Comment (ChrisKujawa) [inline zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/CamundaExporterIT.java:597]: 🔧 we might be able to directly assert the position

File: zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/schema/ElasticsearchEngineClientIT.java
Comment (ChrisKujawa) [inline zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/schema/ElasticsearchEngineClientIT.java:369]: 👍🏼

File: zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/schema/ElasticsearchEngineClientIT.java
Comment (ChrisKujawa) [inline zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/schema/ElasticsearchEngineClientIT.java:319]: 👍🏼

File: zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/schema/ElasticsearchEngineClientIT.java
Comment (ChrisKujawa) [inline zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/schema/ElasticsearchEngineClientIT.java:344]: 👍🏼

**Review by ChrisKujawa — CHANGES_REQUESTED** [review]
Thanks @EuroLew I have some remarks we should take a look at. Please see below. :)

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/schema/elasticsearch/ElasticsearchEngineClient.java
Comment (EuroLew) [reply zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/schema/elasticsearch/ElasticsearchEngineClient.java:201]: The index should always exist either from the importer having created it, or the exporter creates it during the `open` phase.

File: webapps-schema/src/main/java/io/camunda/webapps/schema/descriptors/ZeebeESConstants.java
Comment (EuroLew) [reply webapps-schema/src/main/java/io/camunda/webapps/schema/descriptors/ZeebeESConstants.java:8]: The importers uses these enums to write into the import position index, and so for the exporter to read from the index to know which record reader is done the enums are easiest to use without introducing magic strings because for example.

I think I will take the approach where I avoid moving it and create my own import value type -> index name mapping

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/CamundaExporter.java
Comment (EuroLew) [reply zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/CamundaExporter.java:232]: I didn't think there was an issue with making it configurable, but to avoid any possible issues with this field it's probably better to be not configurable

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/schema/elasticsearch/ElasticsearchEngineClient.java
Comment (EuroLew) [reply zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/schema/elasticsearch/ElasticsearchEngineClient.java:211]: I don't think so, if we also match the completed field then we cannot differentiate between a genuinely empty index and an index where all the documents have completed=false, we could add a query for the amount of documents to differentiate but at that point that's an unnecessary extra query when we could just query all the documents with one request. There will be at most a dozen documents anyway.

File: zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/CamundaExporterIT.java
Comment (EuroLew) [reply zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/CamundaExporterIT.java:597]: Not sure what you mean here I already directly assert the position, do you mean remove the verify call?

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/CamundaExporter.java
Comment (EuroLew) [reply zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/CamundaExporter.java:232]: I don't think there's any reason a user would want to configure this.

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/schema/elasticsearch/ElasticsearchEngineClient.java
Comment (EuroLew) [reply zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/schema/elasticsearch/ElasticsearchEngineClient.java]: I refactored the importers completed logic to also match the completed field

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/utils/SearchEngineClientUtils.java
Comment (EuroLew) [reply zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/utils/SearchEngineClientUtils.java:87]: refactored importers completed logic so this check is not needed

File: zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/CamundaExporterIT.java
Comment (ChrisKujawa) [reply zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/CamundaExporterIT.java:582]: Lol nevermind. I thought we can do `eq(position` but we already do 🤦🏼 Sorry.

File: webapps-schema/src/main/java/io/camunda/webapps/schema/descriptors/ImportValueType.java
Comment (ChrisKujawa) [reply webapps-schema/src/main/java/io/camunda/webapps/schema/descriptors/ImportValueType.java]: Yes I was talking about a new document :)

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/config/ExporterConfiguration.java
Comment (mustafadagher) [inline zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/config/ExporterConfiguration.java:136]: ❓ can we re-use the `bulk.size` here instead of creating a new config value?

File: webapps-schema/src/main/java/io/camunda/webapps/schema/descriptors/ImportValueTypes.java
Comment (mustafadagher) [inline webapps-schema/src/main/java/io/camunda/webapps/schema/descriptors/ImportValueTypes.java:10]: 💭  shall we rename this to `ExportValueType` and reuse it in the `CamundaExporterRecordFilter` instead of the set defined there?

💭 I might be wrong, but I think across the camunda codebase, we use singular rather than plural enum names

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/config/ExporterConfiguration.java
Comment (EuroLew) [reply zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/config/ExporterConfiguration.java:136]: Sure I don't see any harm just wanted to make sure the reasoning was clear

Comment (mustafadagher) [reply zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/config/ExporterConfiguration.java:136]: I understand. My feeling is this is temporary, and it makes sense to stop at the configured bulk.size anyway. We can make it clear through documentation for the 8.7 version

File: webapps-schema/src/main/java/io/camunda/webapps/schema/descriptors/ImportValueTypes.java
Comment (EuroLew) [reply webapps-schema/src/main/java/io/camunda/webapps/schema/descriptors/ImportValueTypes.java:10]: With regards to the first point I don't think so, specifically this refers to all the value types the operate and tasklist importers can handle and we only track if those ones are finished. For example the importers don't handle any `USER` type.

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/config/ExporterConfiguration.java
Comment (EuroLew) [reply zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/config/ExporterConfiguration.java]: Sure I don't see a problem

**Review by mustafadagher — APPROVED** [review]

**Review by ChrisKujawa — CHANGES_REQUESTED** [review]
Thanks @EuroLew I think there is one issue left, regarding merging the value types for the different imports. I think we should take a look at this.

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/schema/elasticsearch/ElasticsearchEngineClient.java
Comment (EuroLew) [reply zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/schema/elasticsearch/ElasticsearchEngineClient.java:193]: Refactored so only one request is made for all documents, therefore no need for a request checking the index is empty

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/CamundaExporter.java
Comment (EuroLew) [reply zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/CamundaExporter.java:253]: Yes, but this call was made before the importers completed being set for tests. If we put this if statement after checking if `importersCompleted` for a fresh install the importsCompleted = true and the job is never scheduled so we can't use the `testcontroller.runScheduledTasks` to manually retry this check.

This is not ideal but logically it doesn't change anything and at worst it's 1 extra check

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/CamundaExporter.java
Comment (EuroLew) [reply zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/CamundaExporter.java:204]: Not sure what you mean by `error loop here for the last record`?

Comment (mustafadagher) [reply zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/CamundaExporter.java:204]: I think Chris was hinting at how we handle an exception in `export(record)`. We don't restart the exporter in case of exception. but rather the `ExporterContainer` will return `false` for this specific record we failed at
https://github.com/camunda/camunda/blob/c3b85b4812a7001bf53504bb11789f1e667d9152/zeebe/broker/src/main/java/io/camunda/zeebe/broker/exporter/stream/ExporterContainer.java#L218-L232

And in the `ExporterDirector` we run this method with retry (with a backoff strategy)
https://github.com/camunda/camunda/blob/c3b85b4812a7001bf53504bb11789f1e667d9152/zeebe/broker/src/main/java/io/camunda/zeebe/broker/exporter/stream/ExporterDirector.java#L654-L655

So in practice: we are (error) looping on exporting this last record that caused this exception

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/schema/elasticsearch/ElasticsearchEngineClient.java
Comment (EuroLew) [reply zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/schema/elasticsearch/ElasticsearchEngineClient.java]: Did not realise that operate and tasklist used two different import position indices, that definitely changes the solution

Comment (EuroLew) [reply zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/schema/elasticsearch/ElasticsearchEngineClient.java]: With the addition of https://github.com/camunda/camunda/pull/25616, the new solution where we check all statuses have `completed=true` should work.

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/CamundaExporter.java
Comment (mustafadagher) [inline zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/CamundaExporter.java:210]: 🔧 this is kind of a nitpicking: but I'm not sure if we should log the whole record. if we need to point to it we shall log the record position.

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/CamundaExporter.java
Comment (mustafadagher) [inline zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/CamundaExporter.java:211]: 🔧 also, is this an error? or more of a warning?

**Comment by mustafadagher** [comment]
`CamundaExporterIT.shouldCreateHarmonizedSchemaEagerlyOnOpen` is failing because we now have an extra index it's not aware of

**Review by ChrisKujawa — CHANGES_REQUESTED** [review]
Hey Euro we are close, I found two more minor things one related to adding more and more records and one regarding tests. Please have a look. If something is unclear please let me know.

In the afternoon I might be able to support, but before 2 my day is quite packed.

**Review by ChrisKujawa — APPROVED** [review]
I think the changes are straightforward so I pre-approve, feel free to merge after fixing the mentioned things above.

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/CamundaExporter.java
Comment (EuroLew) [reply zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/CamundaExporter.java:257]: if it is completed it will simply not reschedule and the task will stop

File: zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/CamundaExporterIT.java
Comment (EuroLew) [reply zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/CamundaExporterIT.java:554]: I added the test the failing when we reach the bulk size, not sure the second test is needed as exporting is idempotent, also I'm not sure why it would write twice? Each `writer.addRecord` call will only add one set of cached entities no matter how many times you call it for a record, and if a record has `export` called on it and succeeds the retry logic shouldn't kick in.

**Pull request merged**
